DM-3473: Ensure that PageBuilder links use appropriate target and styling

camillevilla opened the pull request:
### JIRA issue link
[DM-3473](https://agile6.atlassian.net/browse/DM-3473?atlOrigin=eyJpIjoiYzgyYTc0YjNjZTI1NDI3ZGE3ZjBhYjQ3YTlkMjYxMzgiLCJwIjoiaiJ9)

## Description - what does this code do?
This code handles accessibility best practices for Page Builder links. 
- Remove "Open link in" option from TinyMCE editor 
  - Remediate legacy data without a db update (e.g.internal links set to open in a new tab; external links set to open in the same tab)
- Identify external links
  - Apply USWDS styling so these links receive an icon
  -  set to open in a new tab
- Identify internal links
  - add `.usa-link` styling so that links get `:hover` and `:visited` styles
  - unset any instances of `target='_blank'`

I looked into a Rails approach to doing this but decided to take the JS route. Hopefully this means we won't need to remember to reimplement accessible link behavior as we work on new page builder components.

## Testing done - how did you test it/steps on how can another person can test it 
### In any environment
1. Log in as an admin
2. Edit a page
3. Add a paragraph component and select some text
4. Click on link icon. There should be no "Open link in" dropdown

### In local dev
1. open `active_admin.js` and comment out `_removeTinyMCEOpenLinkInDropdown();` in line 201. 
2. Log is an an admin
3. Create a new page with the following components:
	- paragraph component
  	  - internal link (marketplace.va.gov). Set to open in new tab. 
  	  - internal link (marketplace.va.gov). Set to open in same tab
  	  - internal link (/about). Set to open in new tab
  	  - internal link (/about). Set to open in same tab
  	  - external link. Set to open in same tab
  	- CTA with an external link
  	- Image with an external link
  	- Downloadable file
  	- Practices
  	- Subpage hyperlink
4. Save and publish the page
5. Visit published page and click on each link. 
  - All internal links should open in the same tab
  - All external links
    - should open in a new tab
    - should have "usa-link--external styling" (e.g. icon) UNLESS they are an image or button

## Screenshots, Gifs, Videos from application (if applicable)
### Before
![link-dropdown-before](https://user-images.githubusercontent.com/5402927/179636429-460b61b2-2bb1-4f34-83e7-319772947f34.png)
![content-before](https://user-images.githubusercontent.com/5402927/179636575-eb8c767c-0743-44e5-bc26-7d8ca1bf668d.png)

### After
![link-dropdown-after](https://user-images.githubusercontent.com/5402927/179636432-5c301c15-f40d-402f-a2e5-0260b18afe84.png)
![Screen Shot 2022-07-15 at 3 13 23 PM](https://user-images.githubusercontent.com/5402927/179636456-8f17c362-f43d-42e1-a242-5c349734a00e.png)
![content-after](https://user-images.githubusercontent.com/5402927/179636579-8bdfbf72-efbc-4687-a0c1-f4a534b54c26.png)

## Link to mock-ups/mock ups (image file if you have it) (if applicable)


## Acceptance criteria
- Make internal links in PB automatically open in same tab
- Make external links in PB open in new tab
- Take “Open link in…” dropdown out

## Definition of done
- [ ] Unit tests written (if applicable)
- [x] e2e/accessibility tests written (if applicable)
- [ ] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating JIRA issue
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

bradjohnson92008 reviewed (commented):
So for me I did not have to comment out line 201 in active_admin.js.  I tested all the different page builder component types with links... again only using '%marketplace.va.gov%' for internal links (localhost not included) and they all worked as expected.  
![image](https://user-images.githubusercontent.com/60527788/180051702-389e19aa-5ab2-487a-98e3-776ed584296e.png)

JoshingYou1 commented on app/assets/javascripts/_page_show.es6:
```diff
@@ -58,11 +58,66 @@
+
+    function isPracticeLink(link){
+        return link.hasClass("dm-practice-link");
+    }
```
Instead of having all of these conditions, I was thinking we could add a class to the container div of each `PageParagraphComponent`, and then just check if the closest `div` has that class. So something like:

`page_show.html.erb`
```
<%
  component = pc.component_type.constantize.find(pc.component_id)
  last_component = @page.page_components.size - 1
  is_paragraph_component = pc.component_type == 'PageParagraphComponent'
%>

<%# Page Header and Paragraph %>
<% if (pc.component_type == 'PageHeaderComponent' || is_paragraph_component) && component.text.present? %>
  <div
    class="margin-bottom-4-important <%= page_narrow_classes %><%= ' margin-bottom-0' if last_component === index %><%= 'paragraph-component' if is_paragraph_component %>"
  >
    <%= component.text.html_safe %>
  </div>
<% end %>
```

`_page_show.es6`
```
if (currentLink.closest('div').hasClass('paragraph-component')){
    currentLink.addClass("usa-link");
}
```

camillevilla replied:
ok! I implemented something pretty close to this so that accordion components are also included.

JoshingYou1 commented on app/assets/javascripts/_page_show.es6:
```diff
@@ -58,11 +58,66 @@
+            // Add USWDS link styling text links only
+            if (!isImageLink(currentLink) && !isButton(currentLink) && !isIcon(currentLink) && !isPracticeLink(currentLink)){
+                currentLink.addClass("usa-link");
+            }
```
refer to the comment above

JoshingYou1 commented on app/assets/javascripts/_page_show.es6:
```diff
@@ -58,11 +58,66 @@
+            // Do not apply visual styling to buttons and links
+            if ( !isImageLink(currentLink) && !isButton(currentLink) ) {
+                currentLink.addClass("usa-link usa-link--external");
+            }
```
see top comment

JoshingYou1 reviewed (changes requested):
Great work on this Camille! One thing I noticed was when I created a link on a page and then visited the page, the link was prepended with the page group of the current page. Here's an example: 

![](https://media.giphy.com/media/OTOAjPPHDOISAobnvA/giphy.gif)

I also wasn't able to see the WISYWIG editor when creating a new page. I think it has to do with the second condition on line `197` of `active_admin.js`:
```
if ($body.hasClass('admin_pages') && $body.hasClass('edit')) {
```

![](https://media.giphy.com/media/4kiDr501AQgBEqlcFw/giphy.gif)

camillevilla commented:
ok @bradjohnson92008 @JoshingYou1! Here are my updates since your review: 
- per Josh's suggestion - use CSS classes and restrict link remediation to Paragraph and Accordion components
- fix up CTA button component so that it accounts for internal links 
- DRY things out a by reusing the Ruby logic in CTA button in image links as well
- account for internal links that begin with dots (these happen when users create internal links via TinyMCE) and add some examples of this to the tests
- add Brad's suggested helper text for external URLs

camillevilla commented:
I also looked into the issue with TinyMCE editor disappearing and it seems to be doing that in `master` as well.
>I also wasn't able to see the WISYWIG editor when creating a new page. 
![](https://media.giphy.com/media/4kiDr501AQgBEqlcFw/giphy.gif)

bradjohnson92008 reviewed (commented):
Great Job Camille!  I was able to test through everything - thanks for making all those new changes.  Everything worked as expected for me.

JoshingYou1 commented on app/assets/javascripts/_page_show.es6:
```diff
@@ -58,11 +58,40 @@
 
+    // Remediate internal links on PageBuilder paragraph and accordion components
+    function remediateInternalLinksTarget(){
+        let intLinks = $('.page-paragraph-component, .page-accordion-component').find("a[href*='marketplace.va.gov'], a[href^='/'],a[href^='.']");
```
Question: Do we have any existing urls that start with `.`? Just curious about the use case for that.

camillevilla replied:
Yeah, it's a weird thing Brad found during testing. If you create an internal URL (`/about`) via TinyMCE editor, it'll create a `../../about` in the markup.

JoshingYou1 commented on app/helpers/application_helper.rb:
```diff
@@ -126,7 +126,7 @@ def get_grid_alignment_css_class(alignment)
 
   def get_link_target_attribute(url)
-    if url.include?(ENV.fetch('HOSTNAME'))
+    if url.include?(ENV.fetch('HOSTNAME')) || url.start_with?('/') || url.start_with?('.')
```
Question: Do we have any existing urls that start with .? Just curious about the use case for that.

JoshingYou1 reviewed (commented):
Nice work Camille! Just left a couple of questions/comments 😄

camillevilla commented:
Ok! I updated the semicolons and added the following to the testing instructions:
```
  	  - internal link (/about). Set to open in new tab
  	  - internal link (/about). Set to open in same tab
```

JoshingYou1 reviewed (approved):
LGTM! Great work Camille!

bradjohnson92008 reviewed (approved):
LGTM - great job Camille
